### PR TITLE
chore: consolidate methods for fetching checkpoints in Python SDK [MLG-810]

### PR DESCRIPTION
**Pull request by azhou-determined**
## Description

We currently have a total of 5 methods on `Trial/Experiment` for fetching checkpoints:
```
Experiment.top_checkpoint
Experiment.top_n_checkpoints

Trial.top_checkpoint
Trial.select_checkpoint
Trial.get_checkpoints
```

This change consolidates them to a standard method on each: `.list_checkpoints`, which takes in sort and order parameters. 

**This PR includes changes to backend REST APIs as well as webui methods for accessing them. **

## Test Plan

```
trial = client.get_trial(id=1)
exp = client.get_experiment(id=1)
```

Verify new methods on `Trial`/`Experiment` work with different parameters:
- [ ] sort by enum
- [ ] sort by string
- [ ] limit
```
exp.list_checkpoints(sort_by=checkpoint.CheckpointSortBy.SEARCHER_METRIC, order_by=checkpoint.CheckpointOrderBy.ASC)
exp.list_checkpoints(sort_by="val_metric_name", order_by=checkpoint.CheckpointOrderBy.ASC)
exp.list_checkpoints(limit=1)
...
trial.list_checkpoints(sort_by=checkpoint.CheckpointSortBy.SEARCHER_METRIC, order_by=checkpoint.CheckpointOrderBy.ASC)
trial.list_checkpoints(sort_by="val_metric_name", order_by=checkpoint.CheckpointOrderBy.ASC)
trial.list_checkpoints(limit=1)
```

## Checklist

- [ ] Changes have been manually QA'd
- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
  See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.
- [ ] Licenses should be included for new code which was copied and/or modified from any external code.

### CODE REVIEW

File: docs/model-dev-guide/model-management/checkpoints.rst
Comment (tara-det-ai) [inline docs/model-dev-guide/model-management/checkpoints.rst:34]: The :class:`~determined.experimental.client.Experiment` class is a reference to an experiment. Within this class, there is a :meth:`~determined.experimental.client.Experiment.top_checkpoint` method.
When called without arguments, this method looks at the experiment configuration's searcher field for the
``metric`` and ``smaller_is_better`` values. These values are used to sort the experiment's
checkpoints by validation performance. As an example, the searcher settings in the following experiment configuration file snippet will result in checkpoints being sorted by the loss metric in ascending order.

File: docs/release-notes.rst
Comment (tara-det-ai) [inline docs/release-notes.rst:724]: it is not clear what has been removed

**Review by tara-det-ai — COMMENTED** [review]
added suggested edit

**Review by mapmeld — APPROVED** [review]
WebUI changes are approved

File: proto/src/determined/api/v1/experiment.proto
Comment (stoksc) [inline proto/src/determined/api/v1/experiment.proto:431]: trying to think of a better name.. 'sort by enum' is strange, 'sort-by enum' makes more but i keep reading it the other way.
```suggestion
    determined.checkpoint.v1.SortBy sortByAttr = 2;
```

File: master/internal/api_experiment.go
Comment (stoksc) [inline master/internal/api_experiment.go:1234]: why remove these lines?

File: master/pkg/protoutils/protoless/less.go
Comment (stoksc) [inline master/pkg/protoutils/protoless/less.go:63]: `ai.Training` check is useless because you already checked `ai.Training.ValidationMetrics`, so we would've panicked already. i think we just need to put some nil checks high. rip golang, i want c#'s / ts's `(ai?.Training?.ValidationMetrics?.AvgMetrics ?? {})[metricName]`

**Review by stoksc — COMMENTED** [review]
go side looks good, just minor comments

**Review by tara-det-ai — APPROVED** [review]
Added suggested edits.

File: master/pkg/protoutils/protoless/less.go
Comment (azhou-determined) [reply master/pkg/protoutils/protoless/less.go]: changed to use protobuf getters

File: proto/src/determined/api/v1/experiment.proto
Comment (azhou-determined) [reply proto/src/determined/api/v1/experiment.proto]: Attr makes sense, done

**Review by stoksc — APPROVED** [review]
Go side lgtm

File: docs/model-dev-guide/model-management/checkpoints.rst
Comment (wes-turner) [reply docs/model-dev-guide/model-management/checkpoints.rst:29]: Hot take: when updating, I think the updater should not only add replacement examples that use `list_checkpoints`, but should also remove references to deprecated methods. No reason a user should ever call those.

**Review by wes-turner — COMMENTED** [review]
I like the `list_checkpoints` interface even more than I expected. I think consolidating Checkpoint sort options into one type was really nice, too. Thanks for taking the extra effort to do it completely.

File: docs/model-dev-guide/model-management/checkpoints.rst
Comment (azhou-determined) [reply docs/model-dev-guide/model-management/checkpoints.rst:28]: i'm not sure if deprecation notices belong here, as most people who are reading this page are probably new users? i will defer to @tara-det-ai though.

File: docs/model-dev-guide/model-management/checkpoints.rst
Comment (azhou-determined) [reply docs/model-dev-guide/model-management/checkpoints.rst]: good idea, updated.

File: harness/determined/common/experimental/checkpoint/_checkpoint.py
Comment (azhou-determined) [reply harness/determined/common/experimental/checkpoint/_checkpoint.py]: > It seems more right to me for the value of a CheckpointSortBy instance to be a bindings.checkpointv1SortBy value rather than a bindings.checkpointv1SortBy instance

yeah, this makes sense. updated.

File: harness/determined/common/experimental/experiment.py
Comment (azhou-determined) [reply harness/determined/common/experimental/experiment.py:371]: discussed offline without strong consensus either way, decision is to keep existing interface and ask for feedback after python-sdk feature is ready.

File: harness/determined/common/experimental/experiment.py
Comment (azhou-determined) [reply harness/determined/common/experimental/experiment.py:379]: done

File: harness/determined/common/experimental/experiment.py
Comment (azhou-determined) [reply harness/determined/common/experimental/experiment.py:371]: added a validation type check.

File: docs/model-dev-guide/model-management/checkpoints.rst
Comment (tara-det-ai) [reply docs/model-dev-guide/model-management/checkpoints.rst:28]: i vote leaving it out as long as there is a release note

File: docs/model-dev-guide/model-management/checkpoints.rst
Comment (tara-det-ai) [inline docs/model-dev-guide/model-management/checkpoints.rst:82]: The snippet below demonstrates how to select the most recent checkpoint for a trial.

- or -
The snippet below demonstrates how to select the latest checkpoint for a trial.

**Review by tara-det-ai — APPROVED** [review]
requested edits

File: docs/model-dev-guide/model-management/checkpoints.rst
Comment (wes-turner) [inline docs/model-dev-guide/model-management/checkpoints.rst:96]: This example no longer belongs in the same `code` box as its predecessors.

File: docs/model-dev-guide/model-management/checkpoints.rst
Comment (wes-turner) [reply docs/model-dev-guide/model-management/checkpoints.rst]: These look really good. Nice work!

File: harness/determined/common/experimental/experiment.py
Comment (wes-turner) [reply harness/determined/common/experimental/experiment.py]: Like it.

File: harness/tests/fixtures/api_responses.py
Comment (wes-turner) [inline harness/tests/fixtures/api_responses.py:54]: Could you walk me through the justification to make this `experiment_checkpoints.json` over `checkpoints.json`? Could the file represent any paginated collection of checkpoints?

File: harness/determined/common/experimental/trial.py
Comment (wes-turner) [reply harness/determined/common/experimental/trial.py:397]: Works!

File: harness/determined/common/experimental/experiment.py
Comment (wes-turner) [reply harness/determined/common/experimental/experiment.py:393]: Nope -- I see this is my mistake. The different interface between `list_checkpoints` and the bindings tripped me up.

File: harness/determined/common/experimental/experiment.py
Comment (wes-turner) [reply harness/determined/common/experimental/experiment.py:371]: Temporizing the type of `sort_by`, but the question about bindings not yet addressed.

File: harness/tests/determined/common/experimental/test_experiment.py
Comment (wes-turner) [inline harness/tests/determined/common/experimental/test_experiment.py:235]: I don't tend to like these kinds of tests. If "am I passing through the right arguments" is tricky, then maybe something else has gone wrong. Already written, it doesn't add much overhead -- I guess the test will have to be updated/deleted whenever a parameter changes in either front end or back end -- so it's ok to leave in if you want to. It's just not the kind of thing I like writing tests for.

If there's some complicated calculation one of the parameters undergoes, nice -- I like testing that, specifically. Or if there's tricky edge casing re: parameters and how bindings are called, that's another good thing to test.

I don't know... maybe this is better as a conversation than as a comment. Either here or slack or voice, could you explain how you see it?

File: harness/tests/determined/common/experimental/test_experiment.py
Comment (wes-turner) [inline harness/tests/determined/common/experimental/test_experiment.py:263]: I think this is better as two tests. Both because "Invalid argument set" is hard for an oncall to parse, and each test done separately would be easier to immediately understand and because this is testing two different behaviors.

File: docs/model-dev-guide/model-management/checkpoints.rst
Comment (azhou-determined) [reply docs/model-dev-guide/model-management/checkpoints.rst]: fixed, nice catch

File: docs/model-dev-guide/model-management/checkpoints.rst
Comment (azhou-determined) [reply docs/model-dev-guide/model-management/checkpoints.rst:53]: it's any metric that the user returns in their training loop's validation step. specifics depend on the which Trial/Core API. i guess it's implied that at this point in reading the docs, you've already done the training part which documents returning metrics.

File: docs/model-dev-guide/model-management/checkpoints.rst
Comment (azhou-determined) [reply docs/model-dev-guide/model-management/checkpoints.rst]: this might be an artifact of us being heavily hyperparameter-search oriented... i think making this change is out of scope for this PR since this default comes from the master and i'm not sure how it'd impact other consumers (web) of the API. but yeah i think it's worth asking product/AMLEs.

File: harness/tests/fixtures/api_responses.py
Comment (azhou-determined) [reply harness/tests/fixtures/api_responses.py:54]: yeah good point, updated

File: harness/determined/common/experimental/experiment.py
Comment (azhou-determined) [reply harness/determined/common/experimental/experiment.py:369]: the backend protobuf mandates that exactly one of the two should be set. so if someone calls bindings. get_GetExperimentCheckpoints directly with both, they'll get an APIException.

File: harness/tests/determined/common/experimental/test_experiment.py
Comment (wes-turner) [reply harness/tests/determined/common/experimental/test_experiment.py:235]: > was testing that bindings receives the correct type, either enum or attr (str) from the caller.

That makes sense to me. What do you think about limiting the test's scope to exactly that? And then once that's done, maybe it makes sense to split it into two (also, maybe not)?

File: docs/model-dev-guide/model-management/checkpoints.rst
Comment (wes-turner) [inline docs/model-dev-guide/model-management/checkpoints.rst:105]: I think you accidentally deleted the `get_checkpoint` example. Since it's still a supported way of querying checkpoints, I think it makes sense to include as an example.

File: harness/tests/determined/common/experimental/test_experiment.py
Comment (wes-turner) [reply harness/tests/determined/common/experimental/test_experiment.py]: Looks good! I like these new tests.

Reminder here to do the same for `test_trial.py` before merging.

File: harness/tests/determined/common/experimental/test_experiment.py
Comment (wes-turner) [inline harness/tests/determined/common/experimental/test_experiment.py:235]: tiny nit: someone running tests or seeing the failure won't know from the test name what the failure was about. `test_list_checkpoints_uses_bindings_sortByMetric_on_str_sort_by` does that a little better. But, like, dang that's a tiny nit, and if you prefer this then works for me (also if so, I'm interested in the reasoning so I can learn something).

File: harness/tests/determined/common/experimental/test_trial.py
Comment (wes-turner) [inline harness/tests/determined/common/experimental/test_trial.py:109]: The fact that `Trial.list_checkpoints` uses `api_responses.sample_get_experiment_checkpoints` tells me that one of:
* The api_responses function is improperly named
* Typing is messed up somewhere in/downstream of `Trial.list_checkpoints`
* There's a bug in this test

File: harness/tests/determined/common/experimental/test_experiment.py
Comment (wes-turner) [inline harness/tests/determined/common/experimental/test_experiment.py:261]: Is this actually used in the test?

File: docs/model-dev-guide/model-management/checkpoints.rst
Comment (wes-turner) [inline docs/model-dev-guide/model-management/checkpoints.rst:100]: There are a few more pointers to `top_checkpoint` in this file that, so long as you're fixing "Querying Checkpoints" should be replaced.

File: docs/model-dev-guide/model-management/checkpoints.rst
Comment (azhou-determined) [reply docs/model-dev-guide/model-management/checkpoints.rst:53]: yeah, the user has to know which metrics they reported when training. i guess that's implied but i can't find a great section to link to either. individual training code implementations are going to depend on what kind of Trial/Core API you're using.

File: harness/tests/determined/common/experimental/test_experiment.py
Comment (azhou-determined) [reply harness/tests/determined/common/experimental/test_experiment.py:261]: yeah, it needs to be mocked otherwise `get_with_offset` will throw up.

File: harness/tests/determined/common/experimental/test_experiment.py
Comment (azhou-determined) [reply harness/tests/determined/common/experimental/test_experiment.py:235]: changed to `test_list_checkpoints_calls_bindings_sortByMetric_with_sort_by_str`

File: harness/tests/determined/common/experimental/test_trial.py
Comment (azhou-determined) [reply harness/tests/determined/common/experimental/test_trial.py]: yeah you're right, the api responses returns an experiment checkpoints response which is incorrect. made another method to return the proper trial checkpoints response.

File: harness/tests/determined/common/experimental/test_experiment.py
Comment (wes-turner) [reply harness/tests/determined/common/experimental/test_experiment.py:261]: This didn't make any sense to me, so I some debugging. It's not `get_with_offset` that's throwing up, it's `mock`. [mock expects](https://docs.python.org/3/library/unittest.mock.html#unittest.mock.Mock.side_effect) a `side_effect` to be:
> a function to be called when the mock is called, an iterable or an exception (class or instance) to be raised.

You can fix this by
```python
mock_bindings.side_eddect = [ckpt_resp]
```
or
```python
mock_bindings.return_value = ckpt_resp
```

File: harness/tests/determined/common/experimental/test_experiment.py
Comment (wes-turner) [reply harness/tests/determined/common/experimental/test_experiment.py]: nice

File: docs/model-dev-guide/model-management/checkpoints.rst
Comment (wes-turner) [reply docs/model-dev-guide/model-management/checkpoints.rst:101]: Could you please search for other instances of the deprecated methods? `grep` turns up a couple for me for `top_checkpoint`: (`model-dev-guide/model-management/model-registry-org.rst` and `reference/python-sdk.rst`)

File: harness/tests/determined/common/experimental/test_experiment.py
Comment (wes-turner) [reply harness/tests/determined/common/experimental/test_experiment.py:261]: Oh! Retroactively realizing that my original comment was ambiguous because it's attached to the entire line and not just the `iter_pages` part.

Comment (azhou-determined) [reply harness/tests/determined/common/experimental/test_experiment.py]: nice, thanks

File: docs/model-dev-guide/model-management/checkpoints.rst
Comment (azhou-determined) [reply docs/model-dev-guide/model-management/checkpoints.rst:51]: not sure what you mean, this is what i see:
```   checkpoint = client.get_experiment(id).list_checkpoints()[0]
   specific_checkpoint = client.get_checkpoint(uuid="uuid-for-checkpoint")```

**Review by wes-turner — APPROVED** [review]
LGTM

**Pull request merged**
